### ISSUE TITLE
Implementation of "Independent Validation" in Test & Scores Widget

### ISSUE THREAD

**Issue by flave95** [biolab/orange3#3307]
Hallo Ladys and Gentleman,

In my bachelor thesis i'm working on an implementation of the new validation method "Independent Validation on Classifier results" from Prof. Dr. von Oertzen and Kim Bommae in Orange Canvas. In short, they both proofed that CV is baised and statistically not independent. ( if you want the paper or the scientific proof, i will provide it seperatly to you)
However, in my implementaion work of the new method I discoverd a problem, which I couldnt solve, caused by lack of coding experience or knowledge about python and programming.

The problem is, that I need to manipulate the results, accordingly to calibrate the validator and classifier for the best possible prediction. This means that my code Splits the data, as the all do but in another manner, and present an output in "self." to the results. The actual calculation work is done by other widgets. But the algorithm needs it "back", to add it to learned data and train new on it and give new outcome as the next little test-split. Through this I want to avoid the use of untested data in training, when the classifier should be tested on it.
In addition to that i want to avoid to code a new skl.IndependentValidation / new splitter and just focused on manipulation of the setup_indicies. 

- Examples for other, already exsisting Classifiers in Orange.canvas **and** the "Class Results"
[
        orange3/Orange/evaluation/testing.py
      ](url)
           
##### Expected behavior

1. Split the input data into 
 -train.data (size=0.1)
 -test.data (size=0,9)
2. Split test.data in n_splits= number of observations which are left in 90% big test.data
3. Train on train.data
4. Predict and test on 1_split of group test.data
 -add used 1_split to train.data and remove (pop(0) ) out of n_splits_test.data
5. **Note whrite or wrong prediction for end result ( accuracy and more =correct/ total number)**
6. **Go to 3. until n_splits_test.data = 0**

##### Actual behavior
`class IndependentValidation(Results):
    """
	Info: Independent Validation Testing
	split data into test_data, train_data.
	split test_data in n_splits = number of observations
	---Loop---
	for each n_split_test_data do:
	train on train_data
	predict and test on 1 observation = (1_split_test_data)
	note whrite or wrong prediction in results (self.results?)
	add used 1_split_test_data to train_data
	(re)move used 1_split_test_data from test_data
    do until test_data=0 / while test_data!=0	
	
	"""
      def __init__(self, data, learners, store_data=False, store_models=False, preprocessor=None,
                callback=None, train_size=.1, test_size=None):
          super().__init__(data, learners=learners, store_data=store_data, store_models=store_models, train_data=train_data,
                     preprocessor=preprocessor, callback=callback, train_size=train_size, test_size=test_size)
      def setup_indices(self, train_data, test_data):
          self.data = data
          self.train_data = train_data
          self.left_over_test_data = test_data
          self.train_data, self.left_over_test_data=train_test_split(data, train_size=0.1, random_state=1)
          self.splitted_test_data=get_n_splits(self.left_over_test_data)

      def predict_and_add_to_train_data(self, used_spit, self.splitted_test_data, self.train_data):
          self.test_data = self.splitted_test_data.pop(0)
          #while len(list(self.splitted_test_data)) != 0 #foreach element (self.test_data)
      for self.test_data in self.splittetd_test_data:  # whlie vs. for
          # lm = linear_model.LinearRegression() # exp.
          # model = lm.fit(self.train_data) # exp.
          #self.test_data=results
          #self.train_data=results
          #predictions = lm.predict(self.test_data) # exp.
          self.train_data.append(self.test_data)

##### Steps to reproduce the behavior
on my research i found some infomation on "cross_val_predict" which "Generate cross-validated estimates for each input data point" but i lack of an idea how to use it or 
how to get my for-loop from 3. to 6. on going and interacting with the "Class Results"

[http://scikit-learn.org/stable/modules/generated/sklearn.model_selection.cross_val_predict.html](url)

I hope you can provide me with some help

Greetings

**Comment by markotoplak** [comment]
I am interested to see the paper on "Independent Validation" to actually understand what are you trying to do. I was looking for it myself and it still seems to be under review.

I understand you would like us to help, but it would help a lot if you explained your problems in a concise and clear manner. I tried but I can not understand your problems.

**Comment by flave95** [comment]
Here is a link to the whole paper 
[https://link.springer.com/article/10.3758%2Fs13428-017-0880-z](url)

The problem, as i see it, is that i dont really know how to define the interaction between the IV (independentvalidation) and the Results, or an output defining instance (here i mean the output which will be displayed in test & scores widget table) which calculates the final scores.

Basically, when Cv trains on a subset and validate on another and afterwards use one subset on which it earlier got trained to test on it, all test.data is in process. It is all tossed to a instance which works with the given test and train data accordingly to a defined strategy. 
As shown in this picture, which discrebes the work of CV K-Folds
[https://cdn-images-1.medium.com/max/800/1*J2B_bcbd1-s1kpWOu_FZrg.png](url)

The IV works contrary to this. Its provides at first the calculation with a rather small data set. Train 10% sized and for exp. Test 1%. After the calculation process it adds the 1% to the train data set of 10% and summes it up to train anew on it and to test again 1 another % of the left test.data. This, as it seems to me, needs a defined manipulation of the interaction between IV and the calculation process. They both have to knew that the have to do the process over and over until all test.data was used piece by piece.

**Comment by janezd** [comment]
Closed due to inactivity.
